lib/docker.ts

```diff
@@ -132,69 +180,110 @@ class Docker {
-            if (!Reflect.has(result.txt, filename)) {
-              result.txt[filename] = extracted.layers[layerName].txt[filename];
-            }
-          }
-          // go over MD5 sums found in this layer
-          for (const filename of Object.keys(extracted.layers[layerName].md5)) {
-            // file was not found in previous layer
-            if (!Reflect.has(result.md5, filename)) {
-              result.md5[filename] = extracted.layers[layerName].md5[filename];
-            }
-          }
-        }
-      }
+  /**
+   * Attempt to perform a static scan
+   * @param extractActions array of pattern, callbacks pairs
+   */
+  public async scanStaticalyIfNeeded(
+    extractActions: ExtractAction[],
+  ): Promise<void> {
+    const size = await this.sizeSafe();
+    if (!size || size > this.GetStaticScanSizeLimit()) {
+      return;
+    }
+    try {
+      await this.extractAndCache(extractActions);
+    } catch (error) {
+      debug(error);
+      throw error;
     }
-    return result;
   }
 
   /**
-   * Saves the docker image as a TAR file to a temporary location and extract
-   * the specified files from it.
-   * @param txtPatterns list of plain text key files paths patterns to extract and return as strings
-   * @param md5Patterns list of binary key files paths patterns to extract and return their MD5 sum
-   * @return list of plain text files and list of MD5 sums
+   * Get file product that was previously retrieved and cached or by using
+   *  runtime method after applying the specified callback
+   * @param filename name of file to retrieve its associated string product
+   * @param searchActionName name of a search action
+   * @param callbacks optional array of callbacks to call when runtime method
+   *  is used to retrieve the file
+   * @returns map of file products by callback name
    */
-  public async extract(
-    txtPatterns: string[],
-    md5Patterns: string[] = [],
-  ): Promise<ExtractedKeyFiles> {
-    return this.save(async (err, imageTarPath) => {
-      if (err) {
-        throw err;
+  public async getFileProduct(
+    filename: string,
+    searchActionName: string,
+    extractCallback?: ExtractCallback,
+  ): Promise<string | Buffer> {
+    if (Reflect.has(this.extractProductsByFilename, filename)) {
+      const callbackProducts = this.extractProductsByFilename[filename];
+      if (Reflect.has(callbackProducts, searchActionName)) {
+        return callbackProducts[searchActionName];
       }
-      return await this.analyze(imageTarPath, txtPatterns, md5Patterns);
-    });
+    }
+    const content = (await this.catSafe(filename)).stdout;
+    return extractCallback
+      ? extractCallback(Buffer.from(content, "utf8"))
+      : content;
+  }
+
+  /**
+   * Retrieve all filenames with products of the specified action name
+   * @param searchActionName name of a search action
+   */
+  public getFileProducts(searchActionName: string = "txt"): FilenameProducts {
+    return Object.assign(
```

Comment:
This function is a bit convoluted - how about:
```suggestion
    return Object.keys(this.extractProductsByFilename).reduce(
      (acc : FilenameProducts, file : string) => {
        const val = this.extractProductsByFilename[file][searchActionName];
        return val ? Object.assign(acc, {[file] : val}): acc;
    }, {});
```
Or did I miss something?

Reply:
@mladkau sure, a much elegant approach, thanks